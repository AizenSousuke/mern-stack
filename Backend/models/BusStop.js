const mongoose = require("mongoose");

const BusStopSchema = mongoose.Schema({
	BusStopCode: {
		type: String,
		required: true,
	},
	RoadName: {
		type: String,
		required: true,
	},
	Description: {
		type: String,
		required: true,
	},
	Location: {
		type: { type: String, default: "Point" },
		coordinates: [Number],
	},
});

BusStopSchema.index({ Location: "2dsphere" });

module.exports = BusStop = mongoose.model("busstop", BusStopSchema);
