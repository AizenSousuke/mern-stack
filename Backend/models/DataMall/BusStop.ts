// const mongoose = require("mongoose");

// /**
//  * Schema from LTA DataMall API
//  */
// const BusStopSchema = mongoose.Schema({
// 	BusStopCode: {
// 		type: String,
// 		required: true,
// 	},
// 	RoadName: {
// 		type: String,
// 		required: true,
// 	},
// 	Description: {
// 		type: String,
// 		required: true,
// 	},
// 	Location: {
// 		type: { type: String, default: "Point" },
// 		coordinates: [Number],
// 	},
// });

// BusStopSchema.index({ Location: "2dsphere" });

// const BusStop = mongoose.model("busstop", BusStopSchema);

// export default BusStop;