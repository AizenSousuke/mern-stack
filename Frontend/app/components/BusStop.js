import React, { useState, useEffect } from "react";
import { View, Text, Button } from "react-native";
import BusDetails from "./BusDetails";
import styles from "../../assets/css/AppStyles";

const BusStop = ({ busStopData }) => {
	// console.log(JSON.stringify(busStopData.Services?.sort(a => a.ServiceNo)));
	return (
		<View style={styles.busStop}>
			{busStopData != null ? busStopData.Services?.sort(a => a.ServiceNo).map((service, key) => {
				return (
					<BusDetails
						key={key}
						busNumber={service.ServiceNo}
						details={service}
					/>
				);
			}) : <Text>No bus stop data found</Text>}
		</View>
	);
};

export default BusStop;
