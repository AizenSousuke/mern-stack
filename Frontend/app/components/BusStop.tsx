import React, { useEffect, useState } from "react";
import { View, Text } from "react-native";
import BusDetails from "./BusDetails";
import AppStyles from "../../assets/css/AppStyles";
import { GetBusStopByCode } from "../api/api";

const BusStop = ({ busStopData }: { busStopData: any }) => {
	const [busStopLocation, setBusStopLocation] = useState({ Location: [] });

	useEffect(() => {
		(async () => {
			GetBusStopByCode(busStopData?.BusStopCode).then((result) => {
				setBusStopLocation({ Location: result.busStop?.Location });
			});
		})();
	}, []);

	return (
		<View style={AppStyles.busStop}>
			{busStopData != null ? (
				busStopData.Services?.sort((a: any) => a.ServiceNo).map(
					(service: any, key: number) => {
						return (
							<BusDetails
								key={key}
								busNumber={service.ServiceNo}
								details={service}
								busStopData={busStopData}
								busStopLocation={busStopLocation}
							/>
						);
					}
				)
			) : (
				<Text>No bus stop data found from LTA's API</Text>
			)}
		</View>
	);
};

export default BusStop;
