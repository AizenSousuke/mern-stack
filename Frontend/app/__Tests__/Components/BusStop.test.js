import React from "react";
import { render } from "@testing-library/react-native";
import BusStop from "../../components/BusStop";

// Mock the module as it is rendering a native element
// See: https://github.com/react-native-maps/react-native-maps/issues/2918
jest.mock("react-native-maps", () => {
	const { View } = require("react-native");
	const importedReact = require("react");
	// const MockMapView = (props) => {
	// 	return <View>{props.children}</View>;
	// };

	var mockReact = importedReact;

	const MockMapView = mockReact.forwardRef(
		({ testID, children, ...props }, ref) => {
			if (ref?.current) {
				ref.current = {
					getMapBoundaries: async () => ({
						northEast: {
							latitude: 2,
							longitude: 2,
						},
						southWest: {
							latitude: 1,
							longitude: 1,
						},
					}),
					getCamera: async () => ({
						center: {
							latitude: 2,
							longitude: 2,
						},
						heading: 1,
						pitch: 1,
						zoom: 1,
						altitude: 1,
					}),
					animateCamera: () => {},
				};
			}

			return (
				<View testID={testID} {...props}>
					{children}
				</View>
			);
		}
	);

	// const MockMarker = (props) => {
	// 	return <View>{props.children}</View>;
	// };

	const MockMarker = mockReact.forwardRef(
		({ testID, children, ...props }, ref) => {
			if (ref?.current) {
				ref.current = {
					redraw: () => {},
				};
			}

			return (
				<View testID={testID} {...props}>
					{children}
				</View>
			);
		}
	);

	return {
		__esModule: true,
		default: MockMapView,
		Marker: MockMarker,
	};
});

describe("BusStop", () => {
	const data = {
		BusStopCode: "83139",
		Services: [
			{
				ServiceNo: "911",
				NextBus: {
					OriginCode: "52009",
					DestinationCode: "84009",
					EstimatedArrival: "2017-06-05T14:55:12+08:00",
					Latitude: "1.3184713333333333",
					Longitude: "103.89202066666667",
					VisitNumber: "1",
					Load: "SEA",
					Feature: "WA",
				},
				NextBus2: {
					OriginCode: "52009",
					DestinationCode: "84009",
					EstimatedArrival: "2017-06-05T14:59:12+08:00",
					Latitude: "1.3184713333333333",
					Longitude: "103.89202066666667",
					VisitNumber: "1",
					Load: "SEA",
					Feature: "WA",
				},
				NextBus3: {},
			},
			{
				ServiceNo: "912",
				NextBus: {
					OriginCode: "52009",
					DestinationCode: "84009",
					EstimatedArrival: "2017-06-05T14:55:12+08:00",
					Latitude: "1.3184713333333333",
					Longitude: "103.89202066666667",
					VisitNumber: "1",
					Load: "SEA",
					Feature: "WA",
				},
				NextBus2: {
					OriginCode: "52009",
					DestinationCode: "84009",
					EstimatedArrival: "2017-06-05T14:59:12+08:00",
					Latitude: "1.3184713333333333",
					Longitude: "103.89202066666667",
					VisitNumber: "1",
					Load: "SEA",
					Feature: "WA",
				},
				NextBus3: {},
			},
		],
	};

	it("renders correctly", () => {
		render(<BusStop busStopData={data} />);
	});

	it("renders 2 bus details", () => {
		const busStop = render(<BusStop busStopData={data} />);
		expect(busStop.toJSON().children.length).toEqual(2);
	});

	it("renders bus details properly as per the snapshot", () => {
		const busStop = render(<BusStop busStopData={data} />).toJSON();
		expect(busStop).toMatchSnapshot();
	});

	it("renders no bus stops", async () => {
		const { container, getByText } = render(<BusStop busStopData={null} />);
		expect(getByText("No bus stop data found from LTA's API")).toBeTruthy();
	});
});
