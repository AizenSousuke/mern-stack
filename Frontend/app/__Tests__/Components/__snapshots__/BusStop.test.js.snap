// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Going out renders 2 bus details properly as per snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
    }
  }
>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flexDirection": "row",
        "padding": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "flex": 0.2,
          },
        ]
      }
    >
      911
    </Text>
    <View
      style={
        Object {
          "flex": 0.8,
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.2}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "green",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "green",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -19.2,
                  },
                  Object {
                    "scaleX": 0.20008,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.2}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "green",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "green",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -19.2,
                  },
                  Object {
                    "scaleX": 0.20008,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.9}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "red",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "red",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -2.3999999999999986,
                  },
                  Object {
                    "scaleX": 0.90001,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "backgroundColor": "white",
        "flexDirection": "row",
        "padding": 5,
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "alignSelf": "center",
            "color": "black",
            "fontWeight": "bold",
            "textAlign": "center",
          },
          Object {
            "flex": 0.2,
          },
        ]
      }
    >
      912
    </Text>
    <View
      style={
        Object {
          "flex": 0.8,
          "flexDirection": "row",
          "justifyContent": "space-around",
        }
      }
    >
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.2}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "green",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "green",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -19.2,
                  },
                  Object {
                    "scaleX": 0.20008,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.2}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "green",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "green",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -19.2,
                  },
                  Object {
                    "scaleX": 0.20008,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
      <View
        style={
          Object {
            "flexDirection": "column",
          }
        }
      >
        <Text
          style={
            Object {
              "fontSize": 10,
              "textAlign": "center",
            }
          }
        >
          
        </Text>
        <View
          animated={true}
          animationConfig={
            Object {
              "bounciness": 0,
            }
          }
          animationType="spring"
          indeterminate={false}
          indeterminateAnimationDuration={1000}
          onLayout={[Function]}
          progress={0.9}
          style={
            Array [
              Object {
                "backgroundColor": undefined,
                "borderColor": "red",
                "borderRadius": 4,
                "borderWidth": 1,
                "overflow": "hidden",
                "width": 50,
              },
              undefined,
            ]
          }
          useNativeDriver={false}
        >
          <View
            style={
              Object {
                "backgroundColor": "red",
                "height": 6,
                "transform": Array [
                  Object {
                    "translateX": 0,
                  },
                  Object {
                    "translateX": -2.3999999999999986,
                  },
                  Object {
                    "scaleX": 0.90001,
                  },
                ],
              }
            }
          />
        </View>
        <Text
          style={
            Object {
              "textAlign": "center",
            }
          }
        >
          Arrived
        </Text>
      </View>
    </View>
  </View>
</View>
`;
